homeassistant/components/powerwall/entity.py

```diff
@@ -43,3 +43,45 @@ def __init__(self, powerwall_data: PowerwallRuntimeData) -> None:
     def data(self) -> PowerwallData:
         """Return the coordinator data."""
         return self.coordinator.data
+
+
+class BatteryEntity(CoordinatorEntity[DataUpdateCoordinator[PowerwallData]]):
+    """Base class for battery entities."""
+
+    _attr_has_entity_name = True
+
+    def __init__(
+        self, powerwall_data: PowerwallRuntimeData, battery: BatteryResponse
+    ) -> None:
+        """Initialize the entity."""
+        base_info = powerwall_data[POWERWALL_BASE_INFO]
+        coordinator = powerwall_data[POWERWALL_COORDINATOR]
+        assert coordinator is not None
+        super().__init__(coordinator)
+        self.serial_number = battery.serial_number
+        self.power_wall = powerwall_data[POWERWALL_API]
+        self.base_unique_id = f"{base_info.gateway_din}_{battery.serial_number}"
+
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, self.base_unique_id)},
+            manufacturer=MANUFACTURER,
+            model=f"{MODEL} ({battery.part_number})",
+            name=f"{base_info.site_info.site_name} {battery.serial_number}",
+            sw_version=base_info.status.version,
+            configuration_url=base_info.url,
+            via_device=(DOMAIN, base_info.gateway_din),
+        )
+
+    @property
+    def battery_data(self) -> BatteryResponse:
+        """Return the coordinator data."""
+        for battery in self.coordinator.data.batteries:
+            if battery.serial_number == self.serial_number:
+                return battery
+
+        assert False
```

Comment:
@bdraco Any advice for me on how to deal with the code coverage report on this line?   I'm not sure how to go about testing this.

Reply:
Can we restructure the data so its stored in a dict by serial number so we don't have to linear search it every time?